lib/widgets/app.dart

```diff
@@ -63,22 +65,52 @@ class ZulipApp extends StatefulWidget {
   /// to be mounted.
   static final navigatorKey = GlobalKey<NavigatorState>();
 
+  /// The [ScaffoldMessengerState] for the app.
+  ///
+  /// This is null during the app's early startup, while [ready] is still false.
+  ///
+  /// For code that exists entirely outside the widget tree and has no natural
+  /// [BuildContext] of its own, this enables controlling snack bars.
+  /// Where a relevant [BuildContext] does exist, prefer using that instead,
+  /// with [ScaffoldMessenger.of].
+  static ScaffoldMessengerState? get scaffoldMessenger {
+    final context = navigatorKey.currentContext;
+    if (context == null) return null;
+    // Not maybeOf; we use MaterialApp, which provides ScaffoldMessenger,
+    // so it's a bug if navigatorKey is mounted somewhere lacking that.
+    return ScaffoldMessenger.of(context);
+  }
+
   /// Reset the state of [ZulipApp] statics, for testing.
   ///
   /// TODO refactor this better, perhaps unify with ZulipBinding
   @visibleForTesting
   static void debugReset() {
     _ready.dispose();
     _ready = ValueNotifier(false);
+    reportErrorToUserInSnackBar = null;
+    reportErrorToUserInDialog = null;
   }
 
   /// A list to pass through to [MaterialApp.navigatorObservers].
   /// Useful in tests.
   final List<NavigatorObserver>? navigatorObservers;
 
+  static void _reportErrorToUserInSnackBar(String message) {
+    scaffoldMessenger?.showSnackBar(SnackBar(content: Text(message)));
+  }
+
+  static void _reportErrorToUserInDialog(String title, String message) {
+    final context = navigatorKey.currentContext;
+    if (context == null) return;
```

Comment:
I wonder when this might happen, and whether we want to log to the console in this case, in case there's a developer watching. (Like we do for the case where `ZulipApp` isn't ready.)

Reply:
Hmm, good question. The guarantee we enforce is that when `_ready` is `true`, `currentContext` is always non-null (see `ZulipApp._declareReady`), so it would be a bug if `context` is `null` here. We should use a `!` here and in the other method instead. I also added an assertion `assert(_ready.value)` just to make the invariant explicit.